[Study] add models to get study (chapter) as JSON and PGN

tom-anders opened the pull request.

tom-anders commented:
Not sure what's up with the unrelated test failing. It does not fail when I run it on my machine

tom-anders commented on lib/src/model/study/study.dart:
```diff
@@ -1,11 +1,115 @@
+}
+
+@Freezed(fromJson: true)
+class StudyFeatures with _$StudyFeatures {
```
oh yeah that was an oversight, it's part of `Study` now

tom-anders commented on lib/src/model/study/study.dart:
```diff
@@ -1,11 +1,115 @@
+}
+
+@Freezed(fromJson: true)
+class StudyChapterFeatures with _$StudyChapterFeatures {
```
Done, I had to use `bool?` instead of `bool` then though, since the JSON response seems to not always actually contain these fields (and with a simple record we can't use `@Default`

tom-anders replied:
ah, solved this via `@JsonKey(fromJson: ...)`

veloce reviewed (commented):
Thanks for the changes!

I think now it would be nice to add repository tests, as we have in other repositories. I know it is usually tested end-to-end in a screen test, but here we don't have the screen yet, and I find it interesting to see some examples of the server responses in the tests.

tom-anders commented:
Ok, added tests with a real world study (https://lichess.org/study/JbWtuaeK/7OJXp679)

Also added support for hints and deviation comments (these are not part of the PGN, so we have to parse a subset of `treeParts` for them)

veloce commented on lib/src/model/study/study.dart:
```diff
@@ -9,27 +10,68 @@ import 'package:lichess_mobile/src/model/user/user.dart';
     required IList<String> topics,
     required IList<StudyChapterMeta> chapters,
     required StudyChapter chapter,
+    // Hints to display in "gamebook"/"interactive" mode
```
triple slash :)

tom-anders replied:
done, hope I will stop forgetting this at some point :D

veloce commented on lib/src/model/study/study.dart:
```diff
@@ -9,27 +10,68 @@ import 'package:lichess_mobile/src/model/user/user.dart';
+    // Comment to display when deviating from the mainline in "gamebook" mode
+    // (i.e. when making a wrong move).
+    // Index corresponds to the current ply.
+    required IList<String?> deviationComments,
```
triple slash

tom-anders commented:
(rebased against main)
